File: nprs_app/android/RecognitionSystem/src/main/java/com/nprs/recsystem/image_processing/Convolution.java
package com.nprs.recsystem.image_processing;

import com.nprs.recsystem.common.image.Color;
import com.nprs.recsystem.common.image.Image;
import com.nprs.recsystem.image_processing.filters.Mask;

public class Convolution {
    public static Color convolve(Mask mask, Image image, int x, int y) {
        int maskSize = mask.size();
        int maskMid = mask.size() / 2;
        Color sum = new Color(image.colorInfo());

        for (int mx = 0; mx < maskSize; mx++) {
            for (int my = 0; my < maskSize; my++) {
                int ix = mx + x - maskMid;
                int iy = mx + y - maskMid;

                if (image.isInBounds(ix, iy)) {
                    for (int i = 0; i < sum.numChannels(); i++) {
                        float pixValue = image.getValue(ix, iy, i);
                        sum.setValue(i, pixValue * mask.value(mx, my));
                    }
                }
            }
        }

        return sum;
    }

    public static Color convolve(Mask mask, Image image, int x, int y, int c) {
        int maskSize = mask.size();
        int maskMid = mask.size() / 2;
        Color sum = new Color(image.colorInfo());

        for (int mx = 0; mx < maskSize; mx++) {
            for (int my = 0; my < maskSize; my++) {
                int ix = mx + x - maskMid;
                int iy = mx + y - maskMid;

                if (image.isInBounds(ix, iy)) {
                    for (int i = 0; i < sum.numChannels(); i++) {
                        float pixValue = image.getValue(ix, iy, i);
                        sum.setValue(i, pixValue * mask.value(mx, my));
                    }
                }
            }
        }

        return sum;
    }
}

File: nprs_lib/sources/recognition_system/src/rec_system/image_processing/filters/Convolution.cpp
#include "Convolution.h"
#include <common/image/Image.h>

namespace nprs {

static float convolvePixel(const Image &src, int x, int y, int c, Matrix<float> const& mask);
    
Image convolveImage(Image const& src, Matrix<float> const& mask) {
    Image result(src.width(), src.height(), src.colorInfo());

    for (int channel = 0; channel < src.colorInfo().numChannels(); channel++) {
        convolveChannel(result, channel, src, channel, mask);
    }

    return result;
}

void convolveChannel(
    Image &outImage,
    int outChannel,
    const Image &inImage,
    int inChannel,
    const Matrix<float> &mask)
{
    for (int x = 0; x < inImage.width(); x++) {
        for (int y = 0; y, y < inImage.height(); y++) {
                outImage(x, y, outChannel) = convolvePixel(inImage, x, y, inChannel, mask);
        }
    }
}

static float convolvePixel(Image const& src, int x, int y, int c, Matrix<float> const& mask) {
    const int maskSize = mask.width();
    const int maskMid = maskSize / 2;
    float res = 0.0f;
    for (int qx = -maskSize / 2; qx <= maskSize / 2; qx++) {
        for (int qy = - maskSize / 2; qy <= maskSize / 2; qy++) {
            int mx = maskMid + qx;
            int my = maskMid + qy;
            int px = x + qx;
            int py = y + qy;

            if (src.isInside(px, py))
                res += src(px, py, c) * mask(mx, my);
        }
    }

    return res;
}

}
